Ensure Proxy of non-constructible object is not constructible fixes #4724

rhuanjl opened the pull request:
A rather rough/quick fix for issue #4724 which I just raised.

Not sure if this is best fix but it seems to work

boingoing commented on lib/Runtime/Library/JavascriptProxy.cpp:
```diff
@@ -2191,6 +2191,11 @@ namespace Js
         varArgs[1] = targetObj;
         if (args.Info.Flags & CallFlags_New)
         {
+            if(!JavascriptOperators::IsConstructor(targetObj))
```
Nitpick: space after if

boingoing commented on lib/Runtime/Library/JavascriptProxy.cpp:
```diff
@@ -2039,7 +2039,7 @@ namespace Js
     {
         PROBE_STACK(GetScriptContext(), Js::Constants::MinStackDefault);
-
+        
```
Nitpick: unnecessary whitespace

jackhorton replied:
Turbo-nit: added spaces.

boingoing commented on test/es6/proxybugs.js:
```diff
@@ -378,11 +378,25 @@ var tests = [
+    {
+        name: "Assertion validation : proxy of arrow function should not be constructable",
+        body() {
+            let obj = new Proxy(()=>{}, {construct: ()=>({})});
```
We should test a few more cases like non-constructible builtins (Math.abs, eval, parseInt, etc), 

class members:
```js
class a { 
  static foo() { } 
  bar() { } 
};
new a.foo();
new (new a()).bar();
```
object literal methods:
```js
var obj = { foo() { } };
new obj.foo();
```

jackhorton commented:
@Cellule is this OSX WASM issue known?

rhuanjl commented:
@boingoing points addressed and updated - note I added a lot of test cases so I made a new file for them.

Also the wasm OS X test hasn’t failed after I re-pushed considering that this change is nothing to do with wasm assume it’s just a flaky/slow test.

boingoing commented on test/es6/proxyconstruction.js:
```diff
@@ -0,0 +1,150 @@
+        }
+    },
+    {
+        name: "Proxy of static class member function should be constructable",
```
nitpick: title says this should be constructable but we mean it should "not be" constructable, right?

boingoing commented on test/es6/proxyconstruction.js:
```diff
@@ -0,0 +1,150 @@
+        }
+    },
+    {
+        name: "Proxy of class member function should be constructable",
```
nitpick: as above

boingoing commented:
@rhuanjl Thanks for expanding on the test cases, these are great. Please rebase on top of master and I can help land this one.

rhuanjl commented:
@boingoing it got out of date after I rebased so I rebased again, please let me know if you need me to do anything else.

boingoing commented:
@rhuanjl This looks good to me, I have it ready to go from our internal-side as well. Just fighting with a piece of internal infrastructure and I'll pull the change.

boingoing commented:
@rhuanjl Got it. Thanks very much for your contribution!
